streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -72,6 +83,27 @@ public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(final Ap
         return assignments;
     }
 
+    /**
+     * Assign standby tasks to KafkaStreams clients according to the default logic.
+     * <p>
+     * If rack-aware client tags are configured, the rack-aware standby task assignor will be used
+     *
+     * @param applicationState        the metadata and other info describing the current application state
+     * @param kafkaStreamsAssignments the current assignment of tasks to KafkaStreams clients
+     *
+     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default standby assignment
+     */
+    public static Map<ProcessId, KafkaStreamsAssignment> defaultStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                      final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        if (!applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty()) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
```

Comment:
I'm guessing this represents the middle conditional in the `StandbyTaskAssignorFactory#create` method? I was struggling to understand the point of that because how can you do tag-based rack-aware assignment with no rack-aware assignment tags (ie the first condition)

Then I realized it's actually only ever hit when called from testing code. In the real code, the `rackAwareTaskAssignor` parameter is always null and therefore the middle condition (ie `if (rackAwareTaskAssignor != null && rackAwareTaskAssignor.validClientRack())`) is always false

I have to assume this just exists to make it easier to hook into and/or set up from unit tests, but it's definitely bad practice to have conditions in production code that just exist for test cases. That or it's a bug/regression. Anyways, the tl;dr is I'm pretty sure we should remove these two lines and only have the if/else branches

Reply:
This is driving me crazy, why add so much complexity with these fake "rack" tags if we never use it in non-testing code? I looked up the PR that added this whole thing in the first place ([#14097](https://github.com/apache/kafka/pull/14097)) and the description makes it sound like the "rack" tags are supposed to be a real feature, essentially piggybacking on the older tag-based rack-aware assignment logic. Yet even in that PR, the only thing that actually uses it is from test code. So now I can't tell if it's just an oversight/bug, an unfinished feature that was forgotten about, or really truly just meant to make it easier to write tests. 

Anyways I just need to ask Hao (guy who did the rack-aware optimization KIP) at this point. So in the meantime, until I hear back from him or one of the Confluent folks can answer for him, feel free to ignore all of my comments that relate to this (there are many since this resulted in a lot of branching logic and complexity)